pyg_lib/csrc/ops/matmul.cpp

```diff
@@ -1,33 +1,80 @@
 #include "matmul.h"
-
 #include <ATen/core/dispatch/Dispatcher.h>
 #include <torch/library.h>
+#include <torch/script.h>
 
 namespace pyg {
 namespace ops {
 
+using torch::autograd::AutogradContext;
+using torch::autograd::Variable;
+using torch::autograd::variable_list;
+
 // Performs matrix multiplication across list of elements.
-std::vector<at::Tensor> grouped_matmul(const std::vector<at::Tensor>& input,
-                                       const std::vector<at::Tensor>& other) {
-  // TODO (matthias) Add TensorArg definitions.
-  // TODO (matthias) Add autograd support.
-  // TODO (matthias) Add dispatcher support.
+class GroupedMatmul : public torch::autograd::Function<GroupedMatmul> {
+ public:
   static auto op = c10::Dispatcher::singleton()
                        .findSchemaOrThrow("pyg::grouped_matmul", "")
                        .typed<decltype(grouped_matmul)>();
-  return op.call(input, other);
-}
+  static variable_list forward(AutogradContext* ctx,
+                               std::vector<Variable> input,
+                               std::vector<Variable> other) {
+    auto out = op.call(input, other);
+    ctx->save_for_backward({out, input, other});
+    return out;
+  }
+
+  static variable_list backward(AutogradContext* ctx, variable_list grad_outs) {
+    auto saved = ctx->get_saved_variables();
+    auto out = saved[0];
+    auto input = saved[1];
+    for (size_t i = 0; i < input.size(); ++i)
+      input[i] = input[i].transpose(-2, -1);
+    auto other = saved[2];
+    for (size_t i = 0; i < input.size(); ++i)
+      other[i] = other[i].transpose(-2, -1);
+    auto input_grad = op.call(input, grad_outs);
```

Comment:
I think it is pretty important that we only compute these if `input` requires grad.